CustomElementRegistry.prototype.initialize should upgrade custom elements

rniwa opened the pull request:
#### c13b9f161be7e2cfeb05c08a3041a722516c7fed
<pre>
CustomElementRegistry.prototype.initialize should upgrade custom elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=302534">https://bugs.webkit.org/show_bug.cgi?id=302534</a>

Reviewed by Anne van Kesteren.

Implement <a href="https://github.com/whatwg/html/pull/11913.">https://github.com/whatwg/html/pull/11913.</a> This PR also fixes a few bugs in the existing code, which were revealed by
the newly introduced test.

Test: imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-upgrades.html

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window.js:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-upgrades-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-upgrades.html: Added.

* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::tryToConstructCustomElement): Replaced the bad assertion with a code to get the global object
through the script execution context associated with this custom element interface.

* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::initialize): Implement the new semantics.

* Source/WebCore/dom/Document.cpp:
(WebCore::createUpgradeCandidateElement): Call setUsesNullCustomElementRegistry in the callers of this function (createElementNS
and createElementForBindings).
(WebCore::Document::createElementForBindings):
(WebCore::Document::createElementNS):

Canonical link: <a href="https://commits.webkit.org/303250@main">https://commits.webkit.org/303250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed7bc9ccff4efd77d8e90623b7562681858b9921

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139265 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83627 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4177 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4010 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100707 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134702 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118008 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81489 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/733 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82487 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36135 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141911 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3917 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36712 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109079 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3998 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3447 "Found 1 new test failure: fast/webgpu/present-without-compute-pipeline.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109241 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2979 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114289 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57127 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20492 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3971 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32699 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3803 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67418 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4063 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3931 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
```diff
@@ -120,15 +120,16 @@ RefPtr<Element> JSCustomElementInterface::tryToConstructCustomElement(Document&
 
-    ASSERT(&document == scriptExecutionContext());
-    auto* lexicalGlobalObject = document.globalObject();
+    RefPtr contextDocument = dynamicDowncast<Document>(scriptExecutionContext());
```
This can just be `downcast<>`, no?

rniwa replied:
Indeed. Changed.

rniwa committed:
CustomElementRegistry.prototype.initialize should upgrade custom elements
https://bugs.webkit.org/show_bug.cgi?id=302534

Reviewed by Anne van Kesteren.

Implement https://github.com/whatwg/html/pull/11913. This PR also fixes a few bugs in the existing code, which were revealed by
the newly introduced test.

Test: imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-upgrades.html

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/pseudo-class-defined.window.js:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-upgrades-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-initialize-upgrades.html: Added.

* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::tryToConstructCustomElement): Replaced the bad assertion with a code to get the global object
through the script execution context associated with this custom element interface.

* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::initialize): Implement the new semantics.

* Source/WebCore/dom/Document.cpp:
(WebCore::createUpgradeCandidateElement): Call setUsesNullCustomElementRegistry in the callers of this function (createElementNS
and createElementForBindings).
(WebCore::Document::createElementForBindings):
(WebCore::Document::createElementNS):

Canonical link: https://commits.webkit.org/303250@main
